src/fs/feature/xattr.rs

```diff
@@ -50,59 +50,77 @@ pub enum FollowSymlinks {
 
-    let c_path = match path.to_str().and_then(|s| CString::new(s).ok()) {
-        Some(cstring) => cstring,
+    let c_path = match path.to_str().map(CString::new) {
+        Some(cstring) => match cstring {
+            Ok(cstring) => cstring,
+            Err(e) => {
+                return Err(io::Error::new(io::ErrorKind::Other, e));
+            }
+        },
         None => {
-            return Err(io::Error::new(io::ErrorKind::Other, "Error: path somehow contained a NUL?"));
+            return Err(io::Error::new(io::ErrorKind::Other, "Error: path not convertible to string"));
         }
     };
 
     let bufsize = lister.listxattr_first(&c_path);
-    match bufsize.cmp(&0) {
+    let bufsize = match bufsize.cmp(&0) {
         Ordering::Less     => return Err(io::Error::last_os_error()),
         Ordering::Equal    => return Ok(Vec::new()),
-        Ordering::Greater  => {},
-    }
+        Ordering::Greater  => bufsize as usize,
+    };
 
-    let mut buf = vec![0_u8; bufsize as usize];
-    let err = lister.listxattr_second(&c_path, &mut buf, bufsize);
+    let mut buf = vec![0_u8; bufsize];
 
-    match err.cmp(&0) {
+    match lister.listxattr_second(&c_path, &mut buf, bufsize).cmp(&0) {
         Ordering::Less     => return Err(io::Error::last_os_error()),
         Ordering::Equal    => return Ok(Vec::new()),
         Ordering::Greater  => {},
     }
 
     let mut names = Vec::new();
-    if err > 0 {
-        // End indices of the attribute names
-        // the buffer contains 0-terminated c-strings
-        let idx = buf.iter().enumerate().filter_map(|(i, v)|
-            if *v == 0 { Some(i) } else { None }
-        );
-        let mut start = 0;
-
-        for end in idx {
-            let c_end = end + 1; // end of the c-string (including 0)
-            let size = lister.getxattr(&c_path, &buf[start..c_end]);
-
-            if size > 0 {
-                names.push(Attribute {
-                    name: lister.translate_attribute_name(&buf[start..end]),
-                    size: size as usize,
-                });
+    // End indices of the attribute names
+    // the buffer contains 0-terminated c-strings
+    let idx = buf.iter().enumerate().filter_map(|(i, v)|
+        if *v == 0 { Some(i) } else { None });
+    let mut start = 0;
+
+    for end in idx {
+        let c_attr_name = match CString::new(&buf[start..end]) {
+            Ok(cstring) => cstring,
+            Err(e) => {
+                return Err(io::Error::new(io::ErrorKind::Other, e));
+            }
+        };
+        let size = lister.getxattr_first(&c_path, &c_attr_name);
+
+        if size > 0 {
+            let mut buf_value = vec![0_u8; size as usize];
+            match lister.getxattr_second(&c_path, &c_attr_name, &mut buf_value, size as usize).cmp(&0) {
+                Ordering::Less    => return Err(io::Error::last_os_error()),
+                Ordering::Equal   => {}
+                Ordering::Greater => {}
             }
```

Comment:
I believe this is equivalent and clearer:

```suggestion
            if lister.getxattr_second(&c_path, &c_attr_name, &mut buf_value, size as usize) < 0 {
                return Err(io::Error::last_os_error());
            }
```